File: inverter_hil/+inverterhil/sensorScheduler.m
function [next, due] = sensorScheduler(previous, timeS, input)
%SENSORSCHEDULER Advance synchronized IMU/LWS frame scheduling.
%   Both producers observe the same input state and simulation timestamp.
if nargin < 1 || isempty(previous)
    previous = struct('lastImuS', -Inf, 'lastLwsS', -Inf, ...
        'imuSequence', uint32(0), 'lwsSequence', uint32(0));
end
if nargin < 3 || isempty(input)
    input = struct('imuPeriodS', 0.005, 'lwsPeriodS', 0.010);
end
if ~isscalar(timeS) || ~isfinite(timeS) || timeS < 0
    error('inverterhil:InvalidSensorTime', 'Sensor time must be finite and nonnegative.');
end
required = {'imuPeriodS', 'lwsPeriodS'};
for index = 1:numel(required)
    if ~isfield(input, required{index}) || input.(required{index}) <= 0
        error('inverterhil:InvalidSensorPeriod', 'Sensor periods must be positive.');
    end
end
next = previous;
due = struct('imu', false, 'lws', false, 'timestampS', timeS, ...
    'imuSequence', previous.imuSequence, 'lwsSequence', previous.lwsSequence);
if timeS - previous.lastImuS + eps(timeS + 1) >= input.imuPeriodS
    due.imu = true;
    next.lastImuS = timeS;
    next.imuSequence = previous.imuSequence + uint32(1);
    due.imuSequence = next.imuSequence;
end
if timeS - previous.lastLwsS + eps(timeS + 1) >= input.lwsPeriodS
    due.lws = true;
    next.lastLwsS = timeS;
    next.lwsSequence = previous.lwsSequence + uint32(1);
    due.lwsSequence = next.lwsSequence;
end
end

File: inverter_hil/python/inverterhil/sensor_scheduler.py
import math


def sensorScheduler(previous, timeS, input=None):
    previous = {"lastImuS": -math.inf, "lastLwsS": -math.inf, "imuSequence": 0, "lwsSequence": 0} if not previous else dict(previous)
    input = {"imuPeriodS": 0.005, "lwsPeriodS": 0.010} if not input else input
    if not isinstance(timeS, (int, float)) or isinstance(timeS, bool) or not math.isfinite(timeS) or timeS < 0: raise ValueError("Sensor time must be finite and nonnegative.")
    if any(name not in input or not isinstance(input[name], (int, float)) or input[name] <= 0 for name in ("imuPeriodS", "lwsPeriodS")): raise ValueError("Sensor periods must be positive.")
    nxt = dict(previous); due = {"imu": False, "lws": False, "timestampS": timeS, "imuSequence": previous["imuSequence"], "lwsSequence": previous["lwsSequence"]}
    epsilon = math.ulp(timeS + 1)
    if timeS - previous["lastImuS"] + epsilon >= input["imuPeriodS"]:
        due["imu"] = True; nxt["lastImuS"] = timeS; nxt["imuSequence"] = previous["imuSequence"] + 1; due["imuSequence"] = nxt["imuSequence"]
    if timeS - previous["lastLwsS"] + epsilon >= input["lwsPeriodS"]:
        due["lws"] = True; nxt["lastLwsS"] = timeS; nxt["lwsSequence"] = previous["lwsSequence"] + 1; due["lwsSequence"] = nxt["lwsSequence"]
    return nxt, due
